common/src/services/cipher.service.ts

```diff
@@ -303,22 +309,95 @@ export class CipherService implements CipherServiceAbstraction {
+        } else {
+            this.consoleLogService.info('getAllDecrypted started on main thread');
+            decryptedCiphers = await this.decryptManyCiphers(ciphers);
+        }
+        this.consoleLogService.timeEnd('getAllDecrypted');
+
+        this.decryptedCipherCache = decryptedCiphers.sort(this.getLocaleSortingFunction());
+        return this.decryptedCipherCache;
+    }
+
+    async decryptManyCiphers(ciphers: Cipher[]) {
+        const decryptedCiphers: CipherView[] = [];
+        if (ciphers == null || ciphers.length === 0) {
+            return decryptedCiphers;
+        }
+
+        const hasKey = await this.cryptoService.hasKey();
+        if (!hasKey) {
+            throw new Error('No key.');
+        }
+
+        const promises: any[] = [];
         ciphers.forEach(cipher => {
-            promises.push(cipher.decrypt().then(c => decCiphers.push(c)));
+            promises.push(cipher.decrypt().then(c => decryptedCiphers.push(c)));
         });
 
         await Promise.all(promises);
-        decCiphers.sort(this.getLocaleSortingFunction());
-        this.decryptedCipherCache = decCiphers;
-        return this.decryptedCipherCache;
+        return decryptedCiphers;
+    }
+
+    async decryptManyCiphersWithWorker(workerName: string, cipherData: CipherData[]): Promise<CipherView[]> {
+        if (cipherData == null || cipherData.length === 0) {
+            return null;
+        }
+
+        const cryptoKeys = ConstantsService.cryptoKeys;
+
+        const key = (await this.cryptoService.getKey()).keyB64;
+        const encKey = await this.storageService.get<string>(cryptoKeys.encKey);
+        const orgKeys = await this.storageService.get<string>(cryptoKeys.encOrgKeys);
+        const privateKey = await this.storageService.get<string>(cryptoKeys.encPrivateKey);
+        const storedKeys = {
+            [cryptoKeys.encKey]: encKey,
+            [cryptoKeys.encOrgKeys]: orgKeys,
+            [cryptoKeys.encPrivateKey]: privateKey,
+            [cryptoKeys.key]: key,
+        };
+        const platformUtilsData = {
+            isIE: this.platformUtilsService.isIE(),
+            isSafari: this.platformUtilsService.isSafari(),
+        };
+
+        return new Promise((resolve, reject) => {
+            const worker = this.webWorkerService.create(workerName);
+            worker.postMessage({
+                type: 'decryptManyRequest',
+                ciphers: JSON.stringify(cipherData),
+                keys: JSON.stringify(storedKeys),
+                platformUtilsData: JSON.stringify(platformUtilsData),
+            });
+            worker.addEventListener('message', event => {
+                if (event.data.type !== 'decryptManyResponse') {
+                    return;
+                }
+
+                const decryptedCiphers: CipherView[] = event.data.ciphers.map((v: any) => {
+                    const cipherView = new CipherView();
+                    cipherView.buildFromObj(JSON.parse(v));
```

Comment:
Do we have to `JSON.parse` each object individually? We can't parse once from one array or objects and iterate that?